src/transformers/models/resnet/modeling_tf_resnet.py

```diff
@@ -0,0 +1,542 @@
+        one of `channels_last` (default) or `channels_first`. The ordering of the dimensions in the inputs.
+        `channels_last` corresponds to inputs with shape `(batch, height, width, channels)` while `channels_first`
+        corresponds to inputs with shape `(batch, channels, height, width)`.
+    Input shape:
+      - If `data_format='channels_last'`: 4D tensor with shape `(batch_size, height, width, channels)`.
+      - If `data_format='channels_first'`: 4D tensor with shape `(batch_size, channels, height, width)`.
+    Output shape:
+      - If `data_format='channels_last'`: 4D tensor with shape `(batch_size, pooled_rows, pooled_cols, channels)`.
+      - If `data_format='channels_first'`: 4D tensor with shape `(batch_size, channels, pooled_rows, pooled_cols)`.
+
+    Adapted from [tensorflow-addon's adaptive_pooling.py][
+        https://github.com/tensorflow/addons/blob/8cec33fcaaf1cf90aec7bdd55a0fcdbb251ce5c2/tensorflow_addons/layers/adaptive_pooling.py#L238-L268
+    ]
+    """
+
+    def __init__(self, output_size: Union[int, Iterable[int]], data_format=None, **kwargs) -> None:
+        self.data_format = normalize_data_format(data_format)
+        self.output_size = (output_size, output_size) if isinstance(output_size, int) else tuple(output_size)
+        super().__init__(**kwargs)
+
+    def call(self, inputs: tf.Tensor, *args) -> tf.Tensor:
+        h_bins = self.output_size[0]
+        w_bins = self.output_size[1]
+        if self.data_format == "channels_last":
+            split_cols = tf.split(inputs, h_bins, axis=1)
+            split_cols = tf.stack(split_cols, axis=1)
+            split_rows = tf.split(split_cols, w_bins, axis=3)
+            split_rows = tf.stack(split_rows, axis=3)
+            out_vect = tf.reduce_mean(split_rows, axis=[2, 4])
+        else:
+            split_cols = tf.split(inputs, h_bins, axis=2)
+            split_cols = tf.stack(split_cols, axis=2)
+            split_rows = tf.split(split_cols, w_bins, axis=4)
+            split_rows = tf.stack(split_rows, axis=4)
+            out_vect = tf.reduce_mean(split_rows, axis=[3, 5])
+        return out_vect
+
+    def compute_output_shape(self, input_shape: Iterable[int]) -> tf.TensorShape:
+        input_shape = tf.TensorShape(input_shape).as_list()
+        if self.data_format == "channels_last":
+            shape = tf.TensorShape([input_shape[0], self.output_size[0], self.output_size[1], input_shape[3]])
+        else:
+            shape = tf.TensorShape([input_shape[0], input_shape[1], self.output_size[0], self.output_size[1]])
+        return shape
+
+    def get_config(self):
+        config = {
+            "output_size": self.output_size,
+            "data_format": self.data_format,
+        }
+        base_config = super().get_config()
+        return {**base_config, **config}
+
+
+@add_start_docstrings(
+    "The bare ResNet model outputting raw features without any specific head on top.",
+    RESNET_START_DOCSTRING,
+)
```

Comment:
@amyeroberts we usually also add a `...MainLayer` (which should also be `keras_serializable`) where the components are collated. This main layer is then called from the downstream classes like `...ForImageClassification`, `TFXXXModel`, and so on. See here for an example:

https://github.com/huggingface/transformers/blob/224bde91caff4ccfd12277ab5e9bf97c61e22ee9/src/transformers/models/convnext/modeling_tf_convnext.py#L275

Reply:
Thanks for pointing this out - I wasn't aware. I'll rework it.

Reply:
Done!